subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +328,291 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
+					h.badFreeCellNum[chain][l-1] = 0
 				}
 			}
 		}
 	}
 }
 
-// initFreeCellList initializes the free cell list (i.e., top level cells in each chain).
-func (h *HivedAlgorithm) initFreeCellList() {
-	for chain, ccl := range h.fullCellList {
-		topLevel := CellLevel(len(ccl))
-		h.freeCellList[chain] = NewChainCellList(topLevel - 1)
-		h.freeCellList[chain][topLevel] = make(CellList, len(ccl[topLevel]))
-		copy(h.freeCellList[chain][topLevel], ccl[topLevel])
+// initAPIClusterStatus initiates the status of the physical cluster and the VCs that will be exposed to users.
+func (h *HivedAlgorithm) initAPIClusterStatus() {
+
+	for _, ccl := range h.fullCellList {
+		for _, c := range ccl[CellLevel(len(ccl))] {
+			h.apiClusterStatus.PhysicalCluster = append(h.apiClusterStatus.PhysicalCluster, c.(*PhysicalCell).GetAPIStatus())
+		}
+	}
+	for vc, vcs := range h.vcSchedulers {
+		h.apiClusterStatus.VirtualClusters[vc] = []*api.VirtualCellStatus{}
+		for _, ccl := range vcs.getNonReservedFullCellList() {
+			for _, cl := range ccl {
+				for _, c := range cl {
+					if CellEqual(c, c.(*VirtualCell).GetPreAssignedCell()) {
+						h.apiClusterStatus.VirtualClusters[vc] = append(
+							h.apiClusterStatus.VirtualClusters[vc], c.(*VirtualCell).GetAPIStatus())
+					}
+				}
+			}
+		}
+		for _, ccl := range vcs.getReservedCellList() {
+			for _, c := range ccl[CellLevel(len(ccl))] {
+				h.apiClusterStatus.VirtualClusters[vc] = append(
+					h.apiClusterStatus.VirtualClusters[vc], c.(*VirtualCell).GetAPIStatus())
+			}
+		}
 	}
 }
 
 // initReservations creates static bindings for the reserved cells, and removes the
 // reserved physical cells from the free cell list.
-func (h *HivedAlgorithm) initReservations() {
-	for vc, vcReservation := range h.reservedCells {
+func (h *HivedAlgorithm) initReservations(reservedCells map[api.VirtualClusterName]map[api.ReservationId]*PhysicalCell) {
+	for vcn, vcReservation := range reservedCells {
 		for rid, physical := range vcReservation {
+			h.vcFreeCellNum[vcn][physical.GetChain()][physical.GetLevel()]--
+			h.allVCFreeCellNum[physical.GetChain()][physical.GetLevel()]--
 			h.removeCellFromFreeList(physical)
-			virtualList := h.vcSchedulers[vc].getReservedCellList()[rid]
+			virtualList := h.vcSchedulers[vcn].getReservedCellList()[rid]
 			virtual := virtualList[CellLevel(len(virtualList))][0].(*VirtualCell)
-			virtual.SetPhysicalCell(physical)
-			physical.SetVirtualCell(virtual)
-			klog.Infof("Cells bound: %v and %v (reservation)", virtual.GetName(), physical.GetName())
+			bindCell(physical, virtual)
+		}
+	}
+}
+
+// initBadNodes marks all the physical nodes defined in the config as bad,
+// and wait for K8s's AddNode calls to recognize the healthy nodes.
+func (h *HivedAlgorithm) initBadNodes() {
+	klog.Info("Setting all nodes defined in the config to bad first, " +
+		"and wait for K8s to recognize the healthy nodes (addNode)")
```

Comment:
>recognize [](start = 23, length = 9)

Init all nodes defined in the config to bad, " +
  "and wait for K8s to inform the healthy nodes (addNode) 

?